keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+through CRI APIs or OCI specs will require changes in multiple components
+maintained by different groups/communities when enhancements are planned for
+those parameters. This will make storage enhancements harder and longer to get
+to a "complete" state.
+
+Due to the above constraints, the proposal in the KEP took the approach of
+introducing a new field - `FSDeferralStrategy` - in
+`PersistentVolumeClaimVolumeSource`, `EphemeralVolumeSource` and
+`CSIVolumeSource` structs that conceptually aligns with existing fields like
+`ReadOnly` and specifically instructs how to mount a volume in the context of a
+specific pod. The above constraints leading to the approach in the KEP further
+requires that every CSI plugin directly implement a specific deferral strategy
+supported by a container runtime handler (rather than having the Kubelet
+interact with container runtime handlers to mount and manage file system
+operations). Given limited diversity in the space of general container runtime
+handlers, the number of different strategies a CSI plugin may need to implement
+is expected to be low and mainly geared towards microvms.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+In certain specific scenarios, there are risks associated with enabling deferral
+of file system operations from CSI plugins to container runtime handlers.
+Careful planning is necessa and mitigations for these are enumerated below.
+
+#### Failures after CSI plugin downgrade
+
+If the CSI node plugin is downgraded with respect to it's ability to defer
+file system operations to a container runtime, CSI operations like obtaining
+stats and online expansion of volumes mounted by the container runtime will
+start to fail. The errors should be clearly surfaced to the user (by Kubelet)
+thorugh events on the volume. Recovering from this will require the pod to be
+killed so that the volume gets re-published.
+
+Graceful termination of the pod should work out in most scenarios as the
+downgraded CSI plugin will find it has no mounts to clean up and return success
+(in adherence with the idempotent requirements of multiple invocations of CSI
+APIs). To avoid any failure during NodeUnpublishVolume or leakage of metadata
+files (populated earlier by the CSI plugin in accordance with the file system
+deferral strategy before the downgrade affecting the ability to defer file
+system operations to the container runtime) that are not automatically cleaned
+up the runtime, a node should be drained and cordoned before downgrading the CSI
+node plugin across versions with different capabilities around deferral of file
+system operations to a container runtime handler.
```

Comment:
Since some behavior depends on cluster global CSIDriver and StorageClass annotations / fields:

* In the worst case, can it happen that *all* Pods will end up in a bad state, because CSIDriver / StorageClass annotations do not match the driver version on all / most of the nodes? How will user recover from that situation?
* Would it be possible to set the annotations before driver upgrade / downgrade in a way that all kubelets will fall back to the current CRI behavior, so no such mismatch and manual cleanup will be necessary?

Reply:
Removed the dependence on annotations in CSIDriver and StorageClass.